File: proposed-projects/round-trip-experiment/reverse/20260909T124157Z_4ee67ba2/20260910T071826Z_7d46923a/ported_implementation/gbnf/utils/errors/grammar_parse_error.py
from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f"Failed to parse grammar: {reason}"


def _build_message(grammar: str, pos: int, reason: str) -> str:
    return "\n".join(
        [
            GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
            "",
            *build_error_position(grammar, pos),
        ]
    )


class GrammarParseError(Exception):
    def __init__(self, grammar: str, pos: int, reason: str):
        self.grammar = grammar
        self.pos = pos
        self.reason = reason
        super().__init__(_build_message(grammar, pos, reason))

    def __str__(self) -> str:
        return _build_message(self.grammar, self.pos, self.reason)

    def __repr__(self) -> str:
        return str(self)

    def __eq__(self, other: object) -> bool:
        return str(self) == str(other)

    def __hash__(self) -> int:
        return hash(str(self))

File: packages/gbnf-experiment/data/20260909T022216Z_19e0fe30/ported_implementation/src/utils/errors/grammar-parse-error.ts
import { buildErrorPosition } from './build-error-position.js';

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

const buildMessage = (grammar: string, pos: number, reason: string): string =>
  [
    GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
    '',
    ...buildErrorPosition(grammar, pos),
  ].join('\n');

export class GrammarParseError extends Error {
  grammar: string;
  pos: number;
  reason: string;

  constructor(grammar: string, pos: number, reason: string) {
    super(buildMessage(grammar, pos, reason));
    Object.setPrototypeOf(this, GrammarParseError.prototype);
    this.name = 'GrammarParseError';
    this.grammar = grammar;
    this.pos = pos;
    this.reason = reason;
  }

  toString(): string {
    return buildMessage(this.grammar, this.pos, this.reason);
  }

  equals(other: unknown): boolean {
    return String(this) === String(other);
  }
}
